flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/PaimonEventSink.java

```diff
@@ -63,8 +64,16 @@ public DataStream<Event> addPreWriteTopology(DataStream<Event> dataStream) {
                 .name("Assign Bucket")
                 // All Events after BucketAssignOperator are decorated with BucketWrapper.
                 .partitionCustom(
-                        (bucket, numPartitions) -> bucket % numPartitions,
-                        (event) -> ((BucketWrapper) event).getBucket())
+                        (bucket, numPartitions) -> Math.abs(bucket) % numPartitions,
+                        (event) -> {
+                            if (event instanceof BucketWrapperChangeEvent) {
```

Comment:
It's extremely rare, but `Math.abs(bucket)` still could be negative if `bucket = Integer.MIN_VALUE`. What about using `Math.floorMod(bucket, numPartitions)` to ensure getting a positive mod result here?

Reply:
Done.